### PR TITLE
added autoopen functionality (#477)

### PR DESCRIPTION
**Pull request by alterratz**
implementation of #477 autoopen

f a folder (within the root folder) is named autoload
all entries within this folder are opened automatically

password: the password of the "slave" database
url: the file name of the database which cab be:
   file:///absolute/path.kdbx
   /absolute/path.kdbx
   path.kbdx (which is relative to the current db)

Did not test it on windows or osx but since Qt file
functions have been used I suppose it will work fine.

## Description
iterated over all items in autoload folder in updateTabNameFromDbWidgetSender which is called after a db is opened

also had to change a few lines in DatabaseTabWidget::openDatabase to set passwords for already open databases or ignore databases which have their password set already.

## Motivation and context
implements feature request #477
I use dependent DBs a lot and manually opening them sucks

## How has this been tested?
created a few (8) databases linked them to the master (4) and other "slave" DBs (4) using different url styles and loaded the master db as well as slave DBs
tried it with some of them opened before loading the master as well.

built it with following settings:
cmake -DWITH_XC_AUTOTYPE=ON -DWITH_XC_HTTP=ON -DWITH_XC_YUBIKEY=ON -DWITH_TESTS=ON -DWITH_ASAN=ON -DCMAKE_BUILD_TYPE=Release ..

## Types of changes
- ✅ New feature (non-breaking change which adds functionality)

## Checklist:
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**
- ✅ TBD (when this is the only remaining obstacle for merging) My change requires a change to the documentation and I have updated it accordingly.

### CODE REVIEW

**Comment by droidmonkey** [comment]
Much better!

**Comment by TheZ3ro** [comment]
I would like some check on the file format since `openDatabase` assume it's a kdbx file

File: src/gui/DatabaseTabWidget.cpp [hunk 1]
```diff
@@ -607,6 +613,39 @@ void DatabaseTabWidget::updateTabNameFromDbWidgetSender()
 
     DatabaseWidget* dbWidget = static_cast<DatabaseWidget*>(sender());
     updateTabName(databaseFromDatabaseWidget(dbWidget));
+
+    Database* db = dbWidget->database();
+    Group *autoload = db->rootGroup()->findChildByName("AutoOpen");
+    if (autoload)
+    {
```
Comment (phoerious) [inline src/gui/DatabaseTabWidget.cpp]: Coding style: pointer operator belongs to type, opening brace on same line as if.

File: src/gui/DatabaseTabWidget.cpp [hunk 2]
```diff
@@ -607,6 +613,39 @@ void DatabaseTabWidget::updateTabNameFromDbWidgetSender()
...
+        const DatabaseManagerStruct& dbStruct = m_dbList.value(db);
+        QFileInfo dbpath(dbStruct.canonicalFilePath);
+        QDir dbFolder(dbpath.canonicalPath());
+
+        for (auto entry : autoload->entries()) {
+
```
Comment (phoerious) [inline src/gui/DatabaseTabWidget.cpp]: Remove empty line here.

File: src/gui/DatabaseTabWidget.cpp [hunk 3]
```diff
@@ -607,6 +613,39 @@ void DatabaseTabWidget::updateTabNameFromDbWidgetSender()
...
+            if (entry->url().isEmpty() || entry->password().isEmpty())
+                continue;
+
+            QFileInfo filepath;
+            if (entry->url().startsWith("file:/")) {
```
Comment (phoerious) [inline src/gui/DatabaseTabWidget.cpp]: Shouldn't this be file:// (with a double slash)?

**Comment by alterratz** [comment]
@TheZ3ro :
I am not sure what I shouls check. For instance if i look at main.cc there is also no checking on the command line arguments as they are directly fed to openDatabase.

```    for (int ii=0; ii < args.length(); ii++) {
        QString filename = args[ii];
        if (!filename.isEmpty() && QFile::exists(filename)) {
            QString password;
            if (parser.isSet(pwstdinOption)) {
                static QTextStream in(stdin, QIODevice::ReadOnly);
                password = in.readLine();
            }
            mainWindow.openDatabase(filename, password, parser.value(keyfileOption));
        }

**Comment by TheZ3ro** [comment]
@alterratz I mean a check on the file format, if I try to open a `.txt` file there is something wrong.
Anyway nevermind, the checks are already performed by https://github.com/keepassxreboot/keepassxc/blob/90468e809520f124e8784cf600a90717078d7b45/src/format/KeePass2Reader.cpp#L69 
:wink:

**Comment by phoerious** [comment]
Please update your branch, so I can merge. Thanks!

**Comment by alterratz** [comment]
Updated it to current develop

**Comment by XA21X** [comment]
Finally got it working after realising from the code that "AutoOpen" is the name it's searching for. I must've missed the info about it being the same as KeeAutoExec... Got confused by the "autoload" in the description lol.

**Comment by hrehfeld** [comment]
Is this feature documented somewhere? ( https://duckduckgo.com/?q=keepassxc+auto+open )

**Comment by Emily7777** [comment]
How do we add a key file reference to the auto open vaults entries?

**Comment by droidmonkey** [comment]
Auto-open does not support key files

**Comment by hrehfeld** [comment]
**Bug:** auto open on a file that in turn auto-opens the first file gets into an endless loop. This will probably also work transitively.

**Comment by droidmonkey** [comment]
Don't do that

**Pull request merged**